cpp/oneapi/dal/algo/covariance/test/online_spmd.cpp

```diff
@@ -0,0 +1,131 @@
+#include "oneapi/dal/test/engine/io.hpp"
+
+namespace oneapi::dal::covariance::test {
+
+namespace te = dal::test::engine;
+namespace la = te::linalg;
+namespace cov = oneapi::dal::covariance;
+
+template <typename TestType>
+class covariance_online_spmd_test
+        : public covariance_test<TestType, covariance_online_spmd_test<TestType>> {
+public:
+    using base_t = covariance_test<TestType, covariance_online_spmd_test<TestType>>;
+    using float_t = typename base_t::float_t;
+    using input_t = typename base_t::input_t;
+    using partial_input_t = typename base_t::partial_input_t;
+    using partial_result_t = typename base_t::partial_result_t;
+    using result_t = typename base_t::result_t;
+
+    void set_rank_count(std::int64_t rank_count) {
+        rank_count_ = rank_count;
+    }
+
+    void set_blocks_count(std::int64_t blocks_count) {
+        blocks_count_ = blocks_count;
+    }
+
+    template <typename... Args>
+    result_t finalize_compute_override(Args&&... args) {
+        return this->finalize_compute_via_spmd_threads_and_merge(rank_count_,
+                                                                 std::forward<Args>(args)...);
+    }
+
+    result_t merge_finalize_compute_result_override(const std::vector<result_t>& results) {
+        return results[0];
+    }
+
+    template <typename... Args>
+    std::vector<partial_result_t> split_finalize_compute_input_override(std::int64_t split_count,
+                                                                        Args&&... args) {
+        ONEDAL_ASSERT(split_count == rank_count_);
+        const std::vector<partial_result_t> input{ std::forward<Args>(args)... };
+
+        return input;
+    }
+
+    void online_spmd_general_checks(const te::dataframe& data_fr,
+                                    cov::result_option_id compute_mode,
+                                    const te::table_id& data_table_id) {
+        CAPTURE(static_cast<std::uint64_t>(compute_mode));
+        const table data = data_fr.get_table(this->get_policy(), data_table_id);
+
+        const auto cov_desc = base_t::get_descriptor(compute_mode);
+        std::vector<partial_result_t> partial_results;
+        auto input_table = base_t::template split_table_by_rows<double>(data, rank_count_);
+        for (int64_t i = 0; i < rank_count_; i++) {
+            dal::covariance::partial_compute_result<> partial_result;
+            auto input_table_blocks =
+                base_t::template split_table_by_rows<double>(input_table[i], blocks_count_);
+            for (int64_t j = 0; j < blocks_count_; j++) {
+                partial_result =
+                    this->partial_compute(cov_desc, partial_result, input_table_blocks[j]);
+            }
+            partial_results.push_back(partial_result);
+        }
+        const auto compute_result = this->finalize_compute_override(cov_desc, partial_results);
+
+        base_t::check_compute_result(cov_desc, data, compute_result);
```

Comment:
Does the test check the result on each node? As far as I understand it should as the results should be the same on all nodes.

Reply:
In all of our spmd tests we try to simulate distributed behavior and in other algos just use results[0](in Covariance)